src/compiler/types.rs

```diff
@@ -41,19 +113,84 @@ impl NodeInterface for Node {
             Node::BaseNode(base_node) => base_node.kind(),
             Node::Expression(expression) => expression.kind(),
             Node::Statement(statement) => statement.kind(),
+            Node::SourceFile(source_file) => source_file.kind(),
+        }
+    }
+
+    fn parent(&self) -> Rc<Node> {
+        match self {
+            Node::BaseNode(base_node) => base_node.parent(),
+            Node::Expression(expression) => expression.parent(),
+            Node::Statement(statement) => statement.parent(),
+            Node::SourceFile(source_file) => source_file.parent(),
+        }
+    }
+
+    fn set_parent(&self, parent: Rc<Node>) {
+        match self {
+            Node::BaseNode(base_node) => base_node.set_parent(parent),
+            Node::Expression(expression) => expression.set_parent(parent),
+            Node::Statement(statement) => statement.set_parent(parent),
+            Node::SourceFile(source_file) => source_file.set_parent(parent),
         }
     }
 }
 
 #[derive(Debug)]
 pub struct BaseNode {
     pub kind: SyntaxKind,
+    pub parent: RwLock<Option<Weak<Node>>>,
+    pub pos: AtomicUsize,
+    pub end: AtomicUsize,
+}
+
+impl BaseNode {
+    pub fn new(kind: SyntaxKind, pos: usize, end: usize) -> Self {
+        Self {
+            kind,
+            parent: RwLock::new(None),
+            pos: pos.into(),
+            end: end.into(),
+        }
+    }
 }
 
 impl NodeInterface for BaseNode {
     fn kind(&self) -> SyntaxKind {
         self.kind
     }
+
+    fn parent(&self) -> Rc<Node> {
+        self.parent
+            .read()
+            .unwrap()
+            .clone()
+            .unwrap()
+            .upgrade()
+            .unwrap()
+    }
+
+    fn set_parent(&self, parent: Rc<Node>) {
+        *self.parent.write().unwrap() = Some(Rc::downgrade(&parent));
+    }
+}
+
+impl ReadonlyTextRange for BaseNode {
```

Comment:
This is where it's nice that all of the concrete AST node types eg `PrefixUnaryExpression` are using the same nested `BaseNode` concrete type under the hood for storing the actual `NodeInterface` fields because I only have to implement this atomic stuff once rather than n times